src/RecurringPurchase.sol

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.23;
+
+import {IERC20} from "openzeppelin/token/ERC20/IERC20.sol";
+import {SafeERC20} from "openzeppelin/token/ERC20/utils/SafeERC20.sol";
+
+import {ISwapRouter} from "v3-periphery/interfaces/ISwapRouter.sol";
+
+import {QuarkWallet} from "quark-core/src/QuarkWallet.sol";
+import {QuarkScript} from "quark-core/src/QuarkScript.sol";
+
+/**
+ * @title Recurring Purchase Script
+ * @notice Quark script that performs a swap on a regular interval.
+ * @author Legend Labs, Inc.
+ */
+contract RecurringPurchase is QuarkScript {
+    using SafeERC20 for IERC20;
+
+    error InvalidInput();
+    error PurchaseConditionNotMet();
+
+    /**
+     * @dev Note: This script uses the following storage layout:
+     *         mapping(bytes32 hashedPurchaseConfig => uint256 nextPurchaseTime)
+     *             where hashedPurchaseConfig = keccak256(PurchaseConfig)
+     */
+
+    /// @notice The configuration for a recurring purchase order
+    struct PurchaseConfig {
+        uint256 interval;
+        SwapParams swapParams;
+    }
+
+    /// @notice The set of parameters for performing a swap
+    struct SwapParams {
+        address uniswapRouter;
+        address recipient;
+        address tokenFrom;
+        uint256 amount;
+        uint256 amountInMaximum; // Optional, for "exact out" swaps
+        uint256 amountOutMinimum; // Optional, for "exact in" swaps
```

Comment:
Actually, these won't work for swaps scheduled in the future because they are static. I think we may have to resort to checking the slippage against an oracle (Chainlink or TWAP) as 🌽 suggested.